src/xunit.performance.api/XunitPerformanceHarness.cs

```diff
@@ -81,77 +83,140 @@ public void RunBenchmarks(string assemblyFileName)
         /// If the benchmark scenario has not exited, then it will immediately
         /// stop the associated process, and a TimeoutException will be thrown.
         /// </summary>
-        /// <param name="processStartInfo">The ProcessStartInfo that contains the information that is used to start the benchmark scenario process.</param>
-        /// <param name="preIterationDelegate">The action that will be executed before every benchmark scenario execution.</param>
-        /// <param name="postIterationDelegate">The action that will be executed after every benchmark scenario execution.</param>
+        /// <param name="configuration">ScenarioConfiguration object that defined the scenario execution.</param>
         /// <param name="teardownDelegate">The action that will be executed after running all benchmark scenario iterations.</param>
-        /// <param name="scenarioConfiguration">ScenarioConfiguration object that defined the scenario execution.</param>
-        public void RunScenario(
-            ProcessStartInfo processStartInfo,
-            Action preIterationDelegate,
-            Action postIterationDelegate,
-            Func<ScenarioBenchmark> teardownDelegate,
-            ScenarioConfiguration scenarioConfiguration)
+        public void RunScenario(ScenarioConfiguration configuration, Func<ScenarioBenchmark> teardownDelegate)
         {
-            if (processStartInfo == null)
-                throw new ArgumentNullException($"{nameof(processStartInfo)} cannot be null.");
+            if (configuration == null)
+                throw new ArgumentNullException($"{nameof(configuration)} cannot be null.");
             if (teardownDelegate == null)
                 throw new ArgumentNullException($"{nameof(teardownDelegate)} cannot be null.");
-            if (scenarioConfiguration == null)
-                throw new ArgumentNullException($"{nameof(scenarioConfiguration)} cannot be null.");
 
-            // Make a copy of the user input to avoid potential bugs due to changes behind the scenes.
-            var configuration = new ScenarioConfiguration(scenarioConfiguration);
+            Action<string> OutputFileCallback = (fileName) => {
+                WriteInfoLine($"File saved to: \"{fileName}\"");
+            };
+
+            var scenarioFileName = $"{Configuration.RunId}-{Path.GetFileNameWithoutExtension(configuration.StartInfo.FileName)}";
+            var fileNameWithoutExtension = Path.Combine(OutputDirectory, $"{scenarioFileName}");
 
             for (int i = 0; i < configuration.Iterations; ++i)
             {
-                preIterationDelegate?.Invoke();
+                using (var scenario = new Scenario(configuration))
+                {
+                    ScenarioInfo scenarioInfo;
 
-                // TODO: Start scenario profiling.
+                    configuration.PreIterationDelegate?.Invoke(scenario);
```

Comment:
We should catch and then fail if any custom user code throws when we don't expect it to.